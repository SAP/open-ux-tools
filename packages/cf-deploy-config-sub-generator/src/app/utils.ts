import { isAppStudio } from '@sap-ux/btp-utils';
import { FileName, getMtaPath, getWebappPath } from '@sap-ux/project-access';
import { DeploymentGenerator } from '@sap-ux/deploy-config-generator-shared';
import { ApiHubType, MtaConfig } from '@sap-ux/cf-deploy-config-writer';
import { join } from 'path';
import { bail, ErrorHandler, ERROR_TYPE } from '@sap-ux/deploy-config-generator-shared';
import {
    DEFAULT_MTA_DESTINATION,
    DESTINATION_CHOICE_NONE,
    DESTINATION_CHOICE_DIRECT_SERVICE_BINDING,
    t
} from '../utils';
import type { Manifest } from '@sap-ux/project-access';
import type { Editor } from 'mem-fs-editor';
import type { CfSystemChoice } from '@sap-ux/cf-deploy-config-inquirer';

/**
 * Get the destination choices from API Hub | Local Store | mta.yaml.
 *
 * @param options - the options required for retrieving the questions.
 * @param options.projectRoot - the root path of the project.
 * @param options.isAbapDirectServiceBinding - whether the destination is an ABAP direct service binding.
 * @param options.cfDestination - the Cloud Foundry destination.
 * @param options.isCap - whether the project is a CAP project.
 * @param options.apiHubConfigType - the API Hub configuration.
 * @returns the cf deploy config questions.
 */
export async function getCFChoices({
    projectRoot,
    isAbapDirectServiceBinding,
    cfDestination,
    isCap,
    apiHubConfigType
}: {
    projectRoot: string;
    isAbapDirectServiceBinding: boolean;
    cfDestination: string;
    isCap: boolean;
    apiHubConfigType?: string;
}): Promise<CfSystemChoice[]> {
    let choices: CfSystemChoice[] = [];
    const abapBindingChoice = [
        {
            name: t('cfGen.prompts.abapBinding.name'),
            value: DESTINATION_CHOICE_DIRECT_SERVICE_BINDING
        }
    ];

    // If API Hub Enterprise configuration is used, return the CF destination
    if (apiHubConfigType === ApiHubType.apiHubEnterprise) {
        choices = [
            {
                name: cfDestination,
                value: cfDestination
            }
        ];
    } else {
        const cfChoices =
            !isAbapDirectServiceBinding && isAppStudio()
                ? await getCFSystemChoices(projectRoot, isCap, cfDestination)
                : [];
        choices = isAbapDirectServiceBinding ? abapBindingChoice : cfChoices;
    }
    return choices;
}

/**
 * Generate a systems choice list.
 *
 * @param projectRoot - the root path of the project.
 * @param isCap - whether the project is a CAP project.
 * @param cfDestination - the Cloud Foundry destination.
 * @returns the cf deploy config questions.
 */
async function getCFSystemChoices(
    projectRoot: string,
    isCap: boolean,
    cfDestination?: string
): Promise<CfSystemChoice[]> {
    const choices: CfSystemChoice[] = [];
    try {
        let mtaDestinations: string[] = [];
        // Append mta destinations to support instance based destination flows
        if (isCap) {
            const mtaConfig = await MtaConfig.newInstance(projectRoot);
            mtaDestinations = mtaConfig.getExposedDestinations();
            if (mtaDestinations?.length) {
                // Add default option
                choices.push({
                    name: t('cfGen.prompts.capInstanceBasedDest.name'),
                    value: DEFAULT_MTA_DESTINATION
                });
            }
        } else {
            const mtaResult = await getMtaPath(projectRoot);
            const mtaDir = mtaResult?.mtaPath?.split(FileName.MtaYaml)[0];
            if (mtaDir) {
                const mtaConfig = await MtaConfig.newInstance(mtaDir);
                mtaDestinations = mtaConfig.getExposedDestinations(true);
            }
        }

        // Add MTA destinations
        if (mtaDestinations?.length) {
            // Load additional destinations exposed by the mta.yaml
            mtaDestinations.forEach((dest) => {
                choices.push({
                    name: t('cfGen.prompts.instanceBasedDest.name', { destination: dest }),
                    value: dest
                });
            });
        }
    } catch (error) {
        DeploymentGenerator.logger?.debug(t('cfGen.error.mtaDestinations', { error }));
    }

    if (!cfDestination) {
        choices.splice(0, 0, { name: t('cfGen.prompts.none.name'), value: DESTINATION_CHOICE_NONE });
    }
    return choices;
}

/**
 * Determines the default option for the destination question based on the project environment.
 *
 * @param isAbapDirectServiceBinding - Indicates if ABAP direct service binding is used.
 * @param isBAS - Whether the environment is SAP Business Application Studio (BAS).
 * @param cfDestination - The pre-configured Cloud Foundry destination (if available).
 * @returns {string} The default destination option.
 */
export function destinationQuestionDefaultOption(
    isAbapDirectServiceBinding: boolean,
    isBAS: boolean,
    cfDestination?: string
): string {
    if (!isBAS) {
        return cfDestination || '';
    }
    if (cfDestination) {
        return cfDestination;
    }
    if (isAbapDirectServiceBinding) {
        return DESTINATION_CHOICE_DIRECT_SERVICE_BINDING;
    }
    if (isBAS) {
        return DESTINATION_CHOICE_NONE;
    }
    return '';
}

/**
 * Load the manifest file from the project.
 *
 * @param fs - editor instance
 * @param appPath - path to the project
 * @returns manifest object
 */
export async function loadManifest(fs: Editor, appPath: string): Promise<Manifest | undefined> {
    const manifestPath = join(await getWebappPath(appPath), FileName.Manifest);
    const manifest = fs.readJSON(manifestPath) as unknown as Manifest;

    if (!manifest) {
        bail(ErrorHandler.getErrorMsgFromType(ERROR_TYPE.NO_MANIFEST));
    }

    if (!manifest['sap.app']?.id) {
        bail(ErrorHandler.getErrorMsgFromType(ERROR_TYPE.NO_APP_NAME));
    }

    return manifest;
}
