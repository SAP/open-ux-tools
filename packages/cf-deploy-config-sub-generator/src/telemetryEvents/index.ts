export enum EventName {
    DEPLOY_CONFIG = 'DEPLOY_CONFIG'
}
