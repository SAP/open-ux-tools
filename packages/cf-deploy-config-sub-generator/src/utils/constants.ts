export const API_BUSINESS_HUB_ENTERPRISE_PREFIX = 'ABHE';
export const DEFAULT_MTA_DESTINATION = 'fiori-default-srv-api';
export const DESTINATION_AUTHTYPE_NOTFOUND = 'NotAvailable';
export const DESTINATION_CHOICE_NONE = 'NONE';
export const DESTINATION_CHOICE_DIRECT_SERVICE_BINDING = 'DIRECT_SERVICE_BINDING';
