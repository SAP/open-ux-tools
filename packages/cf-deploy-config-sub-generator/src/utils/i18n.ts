import i18next from 'i18next';
import translations from '../translations/cf-deploy-config-sub-generator.i18n.json';
import { addi18nResourceBundle as addInquirerCommonTexts } from '@sap-ux/inquirer-common';
import type { i18n as i18nNext, TOptions } from 'i18next';

const cfAppRouterGenNs = 'cf-deploy-config-generator';
export const i18n: i18nNext = i18next.createInstance();

/**
 * Initialize i18next with the translations for this module.
 */
export async function initI18n(): Promise<void> {
    await i18n.init({
        lng: 'en',
        fallbackLng: 'en'
    });
    i18n.addResourceBundle('en', cfAppRouterGenNs, translations);

    addInquirerCommonTexts();
}

/**
 * Helper function facading the call to i18next. Unless a namespace option is provided the local namespace will be used.
 *
 * @param key i18n key
 * @param options additional options
 * @returns {string} localized string stored for the given key
 */
export function t(key: string, options?: TOptions): string {
    if (!options?.ns) {
        options = Object.assign(options ?? {}, { ns: cfAppRouterGenNs });
    }
    return i18n.t(key, options);
}

initI18n().catch(() => {
    // Needed for lint
});
