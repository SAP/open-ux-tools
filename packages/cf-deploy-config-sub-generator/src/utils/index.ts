export * from './constants';
export * from './destination';
export * from './i18n';
