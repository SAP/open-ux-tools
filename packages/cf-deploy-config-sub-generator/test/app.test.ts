import fs from 'fs';
import { join } from 'path';
import yeomanTest from 'yeoman-test';
import yaml from 'js-yaml';
import { rimraf } from 'rimraf';
import hasbin from 'hasbin';
import { TestFixture } from './fixtures';
import AppRouterGenerator from '../src/app-router';
import { RouterModuleType } from '@sap-ux/cf-deploy-config-writer';
import { commonChecks } from './utils/checks';
import { initI18n } from '../src/utils';
import { ErrorMessages } from '@sap-ux/deploy-config-generator-shared';
import * as deployConfigGenShared from '@sap-ux/deploy-config-generator-shared';
import * as cfConfigWriter from '@sap-ux/cf-deploy-config-writer';
import * as cfConfigInquirer from '@sap-ux/cf-deploy-config-inquirer';

jest.mock('hasbin', () => ({
    sync: jest.fn()
}));

jest.mock('@sap/mta-lib', () => {
    return {
        Mta: require('./utils/mock-mta').MockMta
    };
});

const hasbinSyncMock = hasbin.sync as jest.MockedFunction<typeof hasbin.sync>;

describe('App router generator tests', () => {
    const targetfolder = join(__dirname, './test-output');
    const testFixture = new TestFixture();
    const appRouterGenPath = join(__dirname, '../src/app-router');

    beforeEach(() => {
        jest.clearAllMocks();
    });

    beforeAll(async () => {
        jest.clearAllMocks();
        await initI18n();
        try {
            fs.mkdirSync(targetfolder);
        } catch {}
    });

    afterAll(() => {
        jest.resetAllMocks();
        rimraf.sync(join(targetfolder, 'sap-ux-test'));
    });

    it('Generate app router project with minimum configuration', async () => {
        const projectPrefix = 'sap-ux-test';
        const appDir = join(targetfolder, projectPrefix);
        hasbinSyncMock.mockReturnValue(true);
        await expect(
            yeomanTest
                .run(AppRouterGenerator, {
                    resolved: appRouterGenPath
                })
                .withOptions({ skipInstall: true })
                .withPrompts({
                    mtaPath: join(targetfolder, '/'),
                    mtaId: 'sap-ux-test',
                    mtaDescription: 'Main MTA configuration for router',
                    routerType: RouterModuleType.Standard,
                    addConnectivityService: false,
                    abapServiceProvider: { label: 'ZZZ_00.0035', service: 'abap-haas' }
                })
        ).resolves.not.toThrow();

        const mtaContent = fs.readFileSync(`${appDir}/mta.yaml`, 'utf-8');
        const mtaConfig = yaml.load(mtaContent);

        const expectMtaContent = testFixture.getContents('sap-ux-test/mta.minimum.yaml');
        const expectMtaConfig = yaml.load(expectMtaContent);
        expect(mtaConfig).toEqual(expectMtaConfig);
        commonChecks(testFixture, appDir);
    });

    it('Generate app router project with connectivity config', async () => {
        hasbinSyncMock.mockReturnValue(true);
        const projectPrefix = 'sap-ux-test';
        const appDir = join(targetfolder, projectPrefix);

        await expect(
            yeomanTest
                .run(AppRouterGenerator, {
                    resolved: appRouterGenPath
                })
                .withOptions({ skipInstall: true })
                .withPrompts({
                    mtaPath: join(targetfolder, '/'),
                    mtaId: 'sap-ux-test',
                    mtaDescription: 'Main MTA configuration for router',
                    routerType: RouterModuleType.Standard,
                    addConnectivityService: true,
                    addABAPServiceBinding: false
                })
        ).resolves.not.toThrow();

        const mtaContent = fs.readFileSync(`${appDir}/mta.yaml`, 'utf-8');
        const mtaConfig = yaml.load(mtaContent);

        const expectMtaContent = testFixture.getContents('sap-ux-test/mta.connectivity.yaml');
        const expectMtaConfig = yaml.load(expectMtaContent);
        expect(mtaConfig).toEqual(expectMtaConfig);

        const xsappContent = fs.readFileSync(`${appDir}/router/xs-app.json`, 'utf-8');
        const xsapp = JSON.parse(xsappContent);
        const expectXsappContent = testFixture.getContents('sap-ux-test/router/xs-app.json');
        const expectXsapp = JSON.parse(expectXsappContent);
        expect(xsapp).toEqual(expectXsapp);
        commonChecks(testFixture, appDir);
    });

    it('Generate app router project with direct abap service config', async () => {
        hasbinSyncMock.mockReturnValue(true);
        const projectPrefix = 'sap-ux-test';
        const appDir = join(targetfolder, projectPrefix);
        await expect(
            yeomanTest
                .run(AppRouterGenerator, {
                    resolved: appRouterGenPath
                })
                .withOptions({ skipInstall: true })
                .withPrompts({
                    mtaPath: join(targetfolder, '/'),
                    mtaId: 'sap-ux-test',
                    mtaDescription: 'Main MTA configuration for router',
                    routerType: RouterModuleType.Standard,
                    addConnectivityService: false,
                    addABAPServiceBinding: true,
                    abapServiceProvider: { label: 'ZZZ_00.0035', service: 'abap-haas' }
                })
        ).resolves.not.toThrow();

        const mtaContent = fs.readFileSync(`${appDir}/mta.yaml`, 'utf-8');
        const mtaConfig = yaml.load(mtaContent);

        const expectMtaContent = testFixture.getContents('sap-ux-test/mta.abapservice.yaml');
        const expectMtaConfig = yaml.load(expectMtaContent);
        expect(mtaConfig).toEqual(expectMtaConfig);

        const xsappContent = fs.readFileSync(`${appDir}/router/xs-app.json`, 'utf-8');
        const xsapp = JSON.parse(xsappContent);
        const expectXsappContent = testFixture.getContents('sap-ux-test/router/xs-app-direct-binding.json');
        const expectXsapp = JSON.parse(expectXsappContent);
        expect(xsapp).toEqual(expectXsapp);
        commonChecks(testFixture, appDir);
    });

    it('Generate app router project with maximum mta config', async () => {
        const projectPrefix = 'sap-ux-test';
        const appDir = join(targetfolder, projectPrefix);

        await expect(
            yeomanTest
                .run(AppRouterGenerator, {
                    resolved: appRouterGenPath
                })
                .withOptions({ skipInstall: true })
                .withPrompts({
                    mtaPath: join(targetfolder, '/'),
                    mtaId: 'sap-ux-test',
                    mtaDescription: 'Main MTA configuration for router',
                    routerType: RouterModuleType.Standard,
                    addConnectivityService: true,
                    addABAPServiceBinding: true,
                    abapServiceProvider: { label: 'ZZZ_00.0035', service: 'abap-haas' }
                })
        ).resolves.not.toThrow();

        const mtaContent = fs.readFileSync(`${appDir}/mta.yaml`, 'utf-8');
        const mtaConfig = yaml.load(mtaContent);
        const expectMtaContent = testFixture.getContents('sap-ux-test/mta.maximum.yaml');
        const expectMtaConfig = yaml.load(expectMtaContent);
        expect(mtaConfig).toEqual(expectMtaConfig);

        const xsappContent = fs.readFileSync(`${appDir}/router/xs-app.json`, 'utf-8');
        const xsapp = JSON.parse(xsappContent);
        const expectXsappContent = testFixture.getContents('sap-ux-test/router/xs-app-direct-binding.json');
        const expectXsapp = JSON.parse(expectXsappContent);
        expect(xsapp).toEqual(expectXsapp);
        commonChecks(testFixture, appDir);
    });

    it('Generate app router project with managed app router', async () => {
        const projectPrefix = 'sap-ux-test';
        const appDir = join(targetfolder, projectPrefix);
        await expect(
            yeomanTest
                .run(AppRouterGenerator, {
                    resolved: appRouterGenPath
                })
                .withOptions({ skipInstall: true })
                .withPrompts({
                    mtaPath: join(targetfolder, '/'),
                    mtaId: 'sap-ux-test',
                    mtaDescription: 'Main MTA configuration for router',
                    mtaVersion: '0.0.1',
                    routerType: RouterModuleType.Managed,
                    addConnectivityService: false,
                    addABAPServiceBinding: false
                })
        ).resolves.not.toThrow();

        const mtaContent = fs.readFileSync(`${appDir}/mta.yaml`, 'utf-8');
        const mtaConfig = yaml.load(mtaContent);
        const expectMtaContent = testFixture.getContents('sap-ux-test/mta.managed.yaml');
        const expectMtaConfig = yaml.load(expectMtaContent);
        expect(mtaConfig).toEqual(expectMtaConfig);
    });

    it('Generate throws error when no mta exe found (CLI)', async () => {
        hasbinSyncMock.mockReturnValue(false);
        // mocking cli behaviour
        jest.spyOn(deployConfigGenShared, 'handleErrorMessage').mockImplementationOnce(() => {
            throw new Error(ErrorMessages.noMtaBin);
        });
        await expect(
            yeomanTest
                .run(AppRouterGenerator, {
                    resolved: appRouterGenPath
                })
                .withOptions({ skipInstall: true, selectedABAPService: true })
        ).rejects.toThrow(ErrorMessages.noMtaBin);
    });

    it('Generate throws error when no mta exe found (VSCODE)', async () => {
        const getAppRouterPromptsSpy = jest.spyOn(cfConfigInquirer, 'getAppRouterPrompts');
        const generateBaseConfigSpy = jest.spyOn(cfConfigWriter, 'generateBaseConfig');
        hasbinSyncMock.mockReturnValue(false);
        // mocking vscode behaviour
        jest.spyOn(deployConfigGenShared, 'handleErrorMessage').mockImplementationOnce(() => {
            console.log(ErrorMessages.noMtaBin);
        });

        const spawnCommandSpy = jest.spyOn(AppRouterGenerator.prototype, 'spawnCommand').mockImplementationOnce(() => {
            Promise.resolve();
        });

        await expect(
            yeomanTest
                .run(AppRouterGenerator, {
                    resolved: appRouterGenPath
                })
                .withOptions({ skipInstall: false })
        ).resolves.not.toThrow();

        expect(getAppRouterPromptsSpy).not.toHaveBeenCalled();
        expect(generateBaseConfigSpy).not.toHaveBeenCalled();
        expect(spawnCommandSpy).toHaveBeenCalled();
    });
});
