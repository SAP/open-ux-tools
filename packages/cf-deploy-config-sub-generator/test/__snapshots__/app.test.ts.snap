// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cloud foundry generator tests Generate CF deployment to an app within a managed app router 4`] = `
[
  {
    "name": "managedApp-destination-service",
    "parameters": {
      "config": {
        "HTML5Runtime_enabled": true,
        "init_data": {
          "instance": {
            "destinations": [
              {
                "Authentication": "NoAuthentication",
                "Name": "ui5",
                "ProxyType": "Internet",
                "Type": "HTTP",
                "URL": "https://ui5.sap.com",
              },
            ],
            "existing_destinations_policy": "update",
          },
        },
        "version": "1.0.0",
      },
      "service": "destination",
      "service-name": "managedApp-destination-service",
      "service-plan": "lite",
    },
    "type": "org.cloudfoundry.managed-service",
  },
  {
    "name": "managedApp_repo_host",
    "parameters": {
      "service": "html5-apps-repo",
      "service-name": "managedApp-html5-srv",
      "service-plan": "app-host",
    },
    "type": "org.cloudfoundry.managed-service",
  },
  {
    "name": "uaa_managedApp",
    "parameters": {
      "path": "./xs-security.json",
      "service": "xsuaa",
      "service-name": "managedApp-xsuaa-srv",
      "service-plan": "application",
    },
    "type": "org.cloudfoundry.managed-service",
  },
]
`;

exports[`Cloud foundry generator tests Generate CF deployment to an app within a managed app router 6`] = `
"{
  "welcomeFile": "/index.html",
  "authenticationMethod": "route",
  "routes": [
    {
      "source": "^/resources/(.*)$",
      "target": "/resources/$1",
      "authenticationType": "none",
      "destination": "ui5"
    },
    {
      "source": "^/test-resources/(.*)$",
      "target": "/test-resources/$1",
      "authenticationType": "none",
      "destination": "ui5"
    },
    {
      "source": "^(.*)$",
      "target": "$1",
      "service": "html5-apps-repo-rt",
      "authenticationType": "xsuaa"
    }
  ]
}
"
`;

exports[`Cloud foundry generator tests Validate app is added and configured for standalone approuter 5`] = `
{
  "devDependencies": {
    "@ui5/cli": "^4.0.16",
    "mbt": "^1.2.29",
    "rimraf": "^5.0.5",
    "ui5-task-zipper": "^3.4.x",
  },
  "scripts": {
    "build:cf": "ui5 build preload --clean-dest --config ui5-deploy.yaml --include-task=generateCachebusterInfo",
    "build:mta": "rimraf resources mta_archives && mbt build",
    "deploy": "fiori cfDeploy",
    "undeploy": "cf undeploy standaloneApp --delete-services --delete-service-keys --delete-service-brokers",
  },
}
`;

exports[`Cloud foundry generator tests Validate app is added and configured for standalone approuter 6`] = `
{
  "authenticationMethod": "route",
  "routes": [
    {
      "authenticationType": "xsuaa",
      "csrfProtection": false,
      "destination": "TestDestination",
      "source": "^/sap/(.*)$",
      "target": "/sap/$1",
    },
    {
      "authenticationType": "none",
      "destination": "ui5",
      "source": "^/resources/(.*)$",
      "target": "/resources/$1",
    },
    {
      "authenticationType": "none",
      "destination": "ui5",
      "source": "^/test-resources/(.*)$",
      "target": "/test-resources/$1",
    },
    {
      "authenticationType": "xsuaa",
      "service": "html5-apps-repo-rt",
      "source": "^(.*)$",
      "target": "$1",
    },
  ],
  "welcomeFile": "/index.html",
}
`;

exports[`Cloud foundry generator tests Validate app is added to an existing managed approuter project with an existing FE app 6`] = `
"{
  "welcomeFile": "/index.html",
  "authenticationMethod": "route",
  "routes": [
    {
      "source": "^/sap/(.*)$",
      "target": "/sap/$1",
      "destination": "testDestination",
      "authenticationType": "xsuaa",
      "csrfProtection": false
    },
    {
      "source": "^/resources/(.*)$",
      "target": "/resources/$1",
      "authenticationType": "none",
      "destination": "ui5"
    },
    {
      "source": "^/test-resources/(.*)$",
      "target": "/test-resources/$1",
      "authenticationType": "none",
      "destination": "ui5"
    },
    {
      "source": "^(.*)$",
      "target": "$1",
      "service": "html5-apps-repo-rt",
      "authenticationType": "xsuaa"
    }
  ]
}
"
`;

exports[`Cloud foundry generator tests Validate new managed approuter is added when there is no existing mta.yaml 2`] = `
{
  "build-parameters": {
    "no-source": true,
  },
  "name": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa-destination-content",
  "parameters": {
    "content": {
      "instance": {
        "destinations": [
          {
            "Name": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa_html_repo_host",
            "ServiceInstanceName": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa-html5-service",
            "ServiceKeyName": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa-repo-host-key",
            "sap.cloud.service": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
          },
          {
            "Authentication": "OAuth2UserTokenExchange",
            "Name": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa_uaa",
            "ServiceInstanceName": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa-xsuaa-service",
            "ServiceKeyName": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa-uaa-key",
            "sap.cloud.service": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
          },
        ],
        "existing_destinations_policy": "update",
      },
    },
  },
  "requires": [
    {
      "name": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa-destination-service",
      "parameters": {
        "content-target": true,
      },
    },
    {
      "name": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa-repo-host",
      "parameters": {
        "service-key": {
          "name": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa-repo-host-key",
        },
      },
    },
    {
      "name": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa-uaa",
      "parameters": {
        "service-key": {
          "name": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa-uaa-key",
        },
      },
    },
  ],
  "type": "com.sap.application.content",
}
`;

exports[`Cloud foundry generator tests Validate new managed approuter is added when there is no existing mta.yaml 3`] = `
{
  "build-parameters": {
    "build-result": "resources",
    "requires": [
      {
        "artifacts": [
          "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa.zip",
        ],
        "name": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
        "target-path": "resources/",
      },
    ],
  },
  "name": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa-app-content",
  "path": ".",
  "requires": [
    {
      "name": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa-repo-host",
      "parameters": {
        "content-target": true,
      },
    },
  ],
  "type": "com.sap.application.content",
}
`;

exports[`Cloud foundry generator tests Validate new managed approuter is added when there is no existing mta.yaml 4`] = `
{
  "build-parameters": {
    "build-result": "dist",
    "builder": "custom",
    "commands": [
      "npm install",
      "npm run build:cf",
    ],
    "supported-platforms": [],
  },
  "name": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
  "path": ".",
  "type": "html5",
}
`;

exports[`Cloud foundry generator tests Validate new managed approuter is added when there is no existing mta.yaml 5`] = `
[
  {
    "name": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa-destination-service",
    "parameters": {
      "config": {
        "HTML5Runtime_enabled": true,
        "init_data": {
          "instance": {
            "destinations": [
              {
                "Authentication": "NoAuthentication",
                "Name": "ui5",
                "ProxyType": "Internet",
                "Type": "HTTP",
                "URL": "https://ui5.sap.com",
              },
            ],
            "existing_destinations_policy": "update",
          },
        },
        "version": "1.0.0",
      },
      "service": "destination",
      "service-name": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa-destination-service",
      "service-plan": "lite",
    },
    "type": "org.cloudfoundry.managed-service",
  },
  {
    "name": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa-uaa",
    "parameters": {
      "path": "./xs-security.json",
      "service": "xsuaa",
      "service-name": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa-xsuaa-service",
      "service-plan": "application",
    },
    "type": "org.cloudfoundry.managed-service",
  },
  {
    "name": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa-repo-host",
    "parameters": {
      "service": "html5-apps-repo",
      "service-name": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa-html5-service",
      "service-plan": "app-host",
    },
    "type": "org.cloudfoundry.managed-service",
  },
]
`;

exports[`Cloud foundry generator tests Validate new managed approuter is added when there is no existing mta.yaml 6`] = `
{
  "description": "Security profile of called application",
  "role-templates": [],
  "scopes": [],
  "tenant-mode": "dedicated",
  "xsappname": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
}
`;

exports[`Cloud foundry generator tests Validate new managed approuter is added when there is no existing mta.yaml 7`] = `
"node_modules/
dist/
.scp/
.env
Makefile*.mta
mta_archives
mta-*
resources
archive.zip
.*_mta_build_tmp"
`;

exports[`Cloud foundry generator tests Validate new managed approuter is added when there is no existing mta.yaml 9`] = `
{
  "authenticationMethod": "route",
  "routes": [
    {
      "authenticationType": "none",
      "csrfProtection": false,
      "destination": "testDestination",
      "source": "^/sap/(.*)$",
      "target": "/sap/$1",
    },
    {
      "authenticationType": "none",
      "destination": "ui5",
      "source": "^/resources/(.*)$",
      "target": "/resources/$1",
    },
    {
      "authenticationType": "none",
      "destination": "ui5",
      "source": "^/test-resources/(.*)$",
      "target": "/test-resources/$1",
    },
    {
      "authenticationType": "xsuaa",
      "service": "html5-apps-repo-rt",
      "source": "^(.*)$",
      "target": "$1",
    },
  ],
  "welcomeFile": "/index.html",
}
`;

exports[`Cloud foundry generator tests Validate new managed approuter is added when there is no existing mta.yaml 10`] = `
{
  "builder": {
    "customTasks": [
      {
        "afterTask": "generateCachebusterInfo",
        "configuration": {
          "additionalFiles": [
            "xs-app.json",
          ],
          "archiveName": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
          "relativePaths": true,
        },
        "name": "ui5-task-zipper",
      },
    ],
    "resources": {
      "excludes": [
        "/test/**",
        "/localService/**",
      ],
    },
  },
  "metadata": {
    "name": "travel",
  },
  "resources": {
    "configuration": {
      "propertiesFileSourceEncoding": "UTF-8",
    },
  },
  "specVersion": "2.4",
  "type": "application",
}
`;
